fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/ContainmentIndexImpl.java

```diff
@@ -87,48 +118,80 @@
             " AND " + TRANSACTION_ID_COLUMN + " = :transactionId AND " + OPERATION_COLUMN + " = 'add'";
 
-    /**
-     * check if a table with the given name exists
-     */
-    private boolean tableIsMissing(final String tableName) throws SQLException {
-        final DatabaseMetaData dbMeta = conn.getMetaData();
-        final ResultSet tables = dbMeta.getTables(null, null, tableName, new String[]{"TABLE"});
-        while (tables.next()) {
-            if (tables.getString("TABLE_NAME").equals(tableName)) {
-                return false;
-            }
-        }
-        return true;
-    }
+    private static final String IS_CHILD_DELETED_IN_TRANSACTION = "SELECT TRUE FROM " + TRANSACTION_OPERATIONS_TABLE +
+            " WHERE " + FEDORA_ID_COLUMN + " = :child AND " + PARENT_COLUMN + " = :parent" +
+            " AND " + TRANSACTION_ID_COLUMN + " = :transactionId AND " + OPERATION_COLUMN + " = 'delete'";
+
+    private static final String ROLLBACK_TRANSACTION = "DELETE FROM " + TRANSACTION_OPERATIONS_TABLE + " WHERE " +
+            " transactionId = :transactionId";
+
+    private static final String COMMIT_ADD_RECORDS = "INSERT INTO " + RESOURCES_TABLE + " ( " + FEDORA_ID_COLUMN + ", "
+            + PARENT_COLUMN + " ) SELECT " + FEDORA_ID_COLUMN + ", " + PARENT_COLUMN + " FROM " +
+            TRANSACTION_OPERATIONS_TABLE + " WHERE " + TRANSACTION_ID_COLUMN + " = :transactionId AND " +
+            OPERATION_COLUMN + " = 'add'";
+
+    private static final String COMMIT_DELETE_RECORDS = "DELETE FROM " + RESOURCES_TABLE + " WHERE EXISTS(" +
+            " SELECT 1 FROM " + TRANSACTION_OPERATIONS_TABLE + " WHERE " + RESOURCES_TABLE + "." + FEDORA_ID_COLUMN +
+            " = " + TRANSACTION_OPERATIONS_TABLE + "." + FEDORA_ID_COLUMN + " AND " + RESOURCES_TABLE + "." +
+            PARENT_COLUMN + " = " + TRANSACTION_OPERATIONS_TABLE + "." + PARENT_COLUMN + " AND " +
+            TRANSACTION_OPERATIONS_TABLE + "." + TRANSACTION_ID_COLUMN + " = :transactionId AND " +
+            TRANSACTION_OPERATIONS_TABLE + "." + OPERATION_COLUMN + " = 'delete')";
+
+    private static final String COMMIT_CLEANUP = "DELETE FROM " + TRANSACTION_OPERATIONS_TABLE + " WHERE " +
+            TRANSACTION_ID_COLUMN + " = :transactionId";
 
     /**
      * Connect to the database
      */
     @PostConstruct
-    private void setup() throws SQLException {
-        conn = dataSource.getConnection();
+    private void setup() {
+        final NamedParameterJdbcTemplate jdbcTemplate = getNamedParameterJdbcTemplate();
+        final MapSqlParameterSource parameterSource = new MapSqlParameterSource();
         // create the tables that don't already exist
-        if (tableIsMissing(RESOURCES_TABLE)) {
-            conn.createStatement().execute(RESOURCES_TABLE_DDL);
+        try {
+            jdbcTemplate.update(RESOURCES_TABLE_DDL, parameterSource);
+        } catch (BadSqlGrammarException e) {
```

Comment:
Couldn't there be other reasons a BadSqlGrammerException is thrown that don't indicate the table already exists? Does the table already exists exception have an error code? This may also vary between db implementations. Can you use a CREATE ... IF NOT EXISTS construct?

Reply:
So there is a wrapped exception, but it is from the database layer which in the test case is h2. If you wire a postgresql or mysql I imagine it will be different.
I can certainly add the h2 exceptions here and once we start wiring in more robust database engines we can extend the `catch`?

Reply:
Alternatively I can re-add the `isTableMissing` function?

Reply:
`CREATE TABLE IF NOT EXISTS` seems most straightforward from my perspective.  Then there won't be any "happy" exceptions to explicitly ignore.